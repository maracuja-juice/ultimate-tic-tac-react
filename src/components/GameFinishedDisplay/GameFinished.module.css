.winnerText {
  margin: 0 0 0.5rem;
}

.restartAlignment {
  display: grid;
  justify-content: center;
  align-items: center;
  margin-bottom: 1rem;
}

@media only screen and (min-width: 505px) {
  .restartAlignment {
    grid-template-columns: auto auto;
  }

  .winnerText {
    margin: 0 1em 0 0;
  }
}

.hidden {
  opacity: 0;
  visibility: hidden;
  transition: all 0.2s ease-out;
}

.visible {
  opacity: 1;
  visibility: visible;
  transition: all 0.2s ease-in;
}
