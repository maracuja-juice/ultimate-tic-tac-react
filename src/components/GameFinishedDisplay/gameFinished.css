.winner-text {
    font-size: 5vmin;
    margin-bottom: 0;
    margin-top: 0;
    margin-right: 1em;
}

.game-finished {
    font-size: 5vmin;
}

.restart-alignment {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-bottom: 1em;
}

.hidden {
    opacity: 0;
    visibility: hidden;
    transition: all 0.2s ease-out;
}

.visible {
    opacity: 1;
    visibility: visible;
    transition: all 0.2s ease-in;
}