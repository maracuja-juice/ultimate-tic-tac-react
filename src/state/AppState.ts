import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
