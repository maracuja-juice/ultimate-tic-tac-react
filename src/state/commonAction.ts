export const RESTART_GAME = 'common/restart-game';

export const restartGame = () => ({
    type: RESTART_GAME
});