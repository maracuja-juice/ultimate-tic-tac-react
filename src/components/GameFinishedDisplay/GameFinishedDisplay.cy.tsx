import * as React from 'react';
import { GameFinishedDisplay } from './GameFinishedDisplay';
import { Winner } from '../../state/AppState';
import { mount } from '@cypress/react';
describe('GameFinishedDisplay', () => {
  it('should change visibility', () => {
    mount(
      <GameFinishedDisplay
        onRestartGame={() => {}}
        isGameFinished={true}
        winner={Winner.Draw}
      />,
    ).then((mountReturn) => {
      cy.findByText(/it's a draw/i).should('be.visible');
      cy.findByRole('button', { name: /play again/i }).should('exist');
      cy.matchImageSnapshot();

      mountReturn.rerender(
        <GameFinishedDisplay
          onRestartGame={() => {}}
          isGameFinished={false}
          winner={Winner.Draw}
        />,
      );

      cy.findByRole('button', { name: /play again/i }).should('not.exist');
      cy.findByText(/it's a draw/i).should('not.be.visible');
      cy.matchImageSnapshot();
    });
  });

  it('should allow clicking restart button', () => {
    mount(
      <GameFinishedDisplay
        onRestartGame={() => {}}
        isGameFinished={true}
        winner={Winner.Cross}
      />,
    );

    const button = cy.findByRole('button', { name: /play again/i });
    button.click();

    cy.matchImageSnapshot();
  });
});
