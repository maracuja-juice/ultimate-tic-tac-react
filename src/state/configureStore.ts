import { AppState } from './AppState';
import { applyMiddleware, combineReducers, createStore } from 'redux';
import logger from 'redux-logger';
import tileReducer from './tiles/tileReducer';
import { composeWithDevTools } from 'redux-devtools-extension';
import gameReducer from './game/gameReducer';
import moveReducer from './moves/moveReducer';
import { enableBatching } from 'redux-batched-actions';
import activeBoardReducer from './activeBoards/activeBoardsReducer';
import createSagaMiddleware from 'redux-saga';
import { all, fork } from 'redux-saga/effects';
import moveValidationSaga from './sagas/MoveValidationSaga';

const rootreducer = combineReducers<AppState>(
    {
        game: gameReducer,
        tiles: tileReducer,
        moves: moveReducer,
        activeBoards: activeBoardReducer
    } );

export function configureStore() {

    const sagaMiddleware = createSagaMiddleware();

    const middleWaresToApply = [
        logger,
        sagaMiddleware
    ];
    const middleware = applyMiddleware( ...middleWaresToApply );
    const store = createStore( enableBatching( rootreducer ), composeWithDevTools(
        middleware
    ) );

    sagaMiddleware.run( rootSaga );

    return store;
}

function* rootSaga() {
    yield all(
        [
            fork( moveValidationSaga ),
        ]
    );
}