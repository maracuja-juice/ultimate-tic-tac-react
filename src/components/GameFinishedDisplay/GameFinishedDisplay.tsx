import * as React from 'react';
import { AppState, Player } from '../../state/AppState';
import { connect } from 'react-redux';
import { XSymbol } from '../symbols/XSymbol';
import { OSymbol } from '../symbols/OSymbol';
import { restartGame } from '../../state/commonAction';
import { Button } from '@rmwc/button';
import './gameFinished.css';

interface GameFinishedDisplayProps {
    isGameFinished: boolean;
    winner: Player;
    onRestartGame: () => void;
}

interface GameFinishedDisplayState {
    winnerClassAttribute: string;
    winnerText: string | JSX.Element;
}

export class GameFinishedDisplay extends React.Component<GameFinishedDisplayProps, GameFinishedDisplayState> {

    constructor( props: GameFinishedDisplayProps ) {
        super( props );

        this.tryRestart = this.tryRestart.bind( this );
        this.getWinnerText = this.getWinnerText.bind( this );

        this.state = {winnerClassAttribute: 'hidden', winnerText: this.getWinnerText( Player.Circle, false )};
    }

    getWinnerText( player: Player, isGameFinished: boolean ) {

        if (isGameFinished) {
            if (player === Player.Circle) {
                return (<><OSymbol className="winner-symbol" shouldAnimate={false}/> wins!</>);
            } else if (player === Player.Cross) {
                return (<><XSymbol className="winner-symbol" shouldAnimate={false}/> wins!</>);
            } else {
                return `It's a draw!`;
            }
        }

        // this will be hidden
        // it's there so that the board doesn't shift down when the finished text is displayed
        return <><XSymbol shouldAnimate={false}/>reservation</>;
    }

    componentWillReceiveProps( nextProps: GameFinishedDisplayProps ) {
        if (nextProps.isGameFinished) {
            this.setState( {
                               winnerClassAttribute: 'visible',
                               winnerText: this.getWinnerText( nextProps.winner, nextProps.isGameFinished )
                           } );
        }
    }

    tryRestart() {
        if (this.props.isGameFinished) {
            // for the ease out to work.
            // otherwise the "reservation" text gets rendered before it's eased out
            this.setState( prevState => {
                return {
                    ...prevState,
                    winnerClassAttribute: 'hidden'
                };
            } );

            this.props.onRestartGame();
        }
    }

    render() {
        const {winnerClassAttribute, winnerText} = this.state;

        const textContainerClass = 'restart-alignment ' + winnerClassAttribute;

        return (
            <div className={textContainerClass}>
                <p className="winner-text">
                    {winnerText}
                </p>
                <Button dense={true} raised={true} onClick={this.tryRestart}>
                    Play Again
                </Button>
            </div>
        );
    }
}

const mapStateToProps = ( state: AppState ) => ({
    isGameFinished: state.game.isFinished,
    winner: state.game.winningPlayer,
});

// tslint:disable-next-line: no-any
const mapDispatchToProps = ( dispatch: any ) => ({
    onRestartGame: () =>
        dispatch( restartGame() )
});

export default connect( mapStateToProps, mapDispatchToProps )( GameFinishedDisplay );
