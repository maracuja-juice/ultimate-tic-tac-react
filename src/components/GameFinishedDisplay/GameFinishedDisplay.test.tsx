import * as React from 'react';
import { configure, shallow } from 'enzyme';
import * as ReactSixteenAdapter from 'enzyme-adapter-react-16';
import { Player } from '../../state/AppState';
import { GameFinishedDisplay } from './GameFinishedDisplay';

configure( {adapter: new ReactSixteenAdapter()} );

describe( 'GameFinished', function () {

    it( 'should match snapshot when draw', () => {
        // tslint:disable:no-empty
        const restart = () => {
        };
        const gameFinishedIndicator = shallow( (
                                                   <GameFinishedDisplay
                                                       onRestartGame={restart}
                                                       isGameFinished={true}
                                                       winner={null!}
                                                   />
                                               ) );

        expect( gameFinishedIndicator ).toMatchSnapshot();
    } );

    it( 'should match snapshot when circle wins', () => {
        // tslint:disable:no-empty
        const restart = () => {
        };
        const gameFinishedIndicator = shallow( (
                                                   <GameFinishedDisplay
                                                       onRestartGame={restart}
                                                       isGameFinished={true}
                                                       winner={Player.Circle}
                                                   />
                                               ) );

        expect( gameFinishedIndicator ).toMatchSnapshot();
    } );

    it( 'should match snapshot when cross wins', () => {
        // tslint:disable:no-empty
        const restart = () => {
        };
        const gameFinishedIndicator = shallow( (
                                                   <GameFinishedDisplay
                                                       onRestartGame={restart}
                                                       isGameFinished={true}
                                                       winner={Player.Cross}
                                                   />
                                               ) );

        expect( gameFinishedIndicator ).toMatchSnapshot();
    } );

    it( 'should match snapshot when no one wins', () => {
        // tslint:disable:no-empty
        const restart = () => {
        };
        const gameFinishedIndicator = shallow( (
                                                   <GameFinishedDisplay
                                                       onRestartGame={restart}
                                                       isGameFinished={false}
                                                       winner={null!}
                                                   />
                                               ) );

        expect( gameFinishedIndicator ).toMatchSnapshot();
    } );
} );