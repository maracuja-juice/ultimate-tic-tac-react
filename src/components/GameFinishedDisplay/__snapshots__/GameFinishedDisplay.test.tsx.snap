// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameFinished should match snapshot when circle wins 1`] = `
<div
  className="restartAlignment visible"
>
  <Typography
    className="winnerText"
    use="headline5"
  >
    <Icon
      className="floatLeft winnerIconContainer"
      icon={
        Object {
          "icon": "o",
          "size": "large",
        }
      }
    />
    <span
      className="floatRight"
    >
      wins!
    </span>
  </Typography>
  <withRipple(Button)
    dense={true}
    onClick={[Function]}
    raised={true}
  >
    Play Again
  </withRipple(Button)>
</div>
`;

exports[`GameFinished should match snapshot when cross wins 1`] = `
<div
  className="restartAlignment visible"
>
  <Typography
    className="winnerText"
    use="headline5"
  >
    <Icon
      className="floatLeft winnerIconContainer"
      icon={
        Object {
          "icon": "x",
          "size": "large",
        }
      }
    />
    <span
      className="floatRight"
    >
      wins!
    </span>
  </Typography>
  <withRipple(Button)
    dense={true}
    onClick={[Function]}
    raised={true}
  >
    Play Again
  </withRipple(Button)>
</div>
`;

exports[`GameFinished should match snapshot when draw 1`] = `
<div
  className="restartAlignment visible"
>
  <Typography
    className="winnerText"
    use="headline5"
  >
    It's a draw!
  </Typography>
  <withRipple(Button)
    dense={true}
    onClick={[Function]}
    raised={true}
  >
    Play Again
  </withRipple(Button)>
</div>
`;

exports[`GameFinished should match snapshot when no one wins 1`] = `
<div
  className="restartAlignment hidden"
>
  <Typography
    className="winnerText"
    use="headline5"
  >
    <Icon
      icon={
        Object {
          "icon": "x",
          "size": "large",
        }
      }
    />
    <span
      className="floatRight"
    >
      reservation
    </span>
  </Typography>
  <withRipple(Button)
    dense={true}
    onClick={[Function]}
    raised={true}
  >
    Play Again
  </withRipple(Button)>
</div>
`;
